File: leetcode/201-300/238-product-of-array-except-self/bench/product_except_self.py
"""Benchmark harness for LeetCode #238 — Product of Array Except Self.

Mirrors product_except_self.kara algorithm-for-algorithm. Committed as the
correctness oracle; not a measured lane.

NOTE on `%`: the sink is negative, and Python's `%` FLOORS while Kara, Rust, C
and Go all TRUNCATE toward zero. Using Python's operator directly would produce
a different sink from the other four for the same correct computation, so
`trunc_mod` below reproduces the truncating semantics.
"""

NP = 8
N = 100000
ITERS = 400


def trunc_mod(a, b):
    r = a % b
    if r != 0 and (a < 0) != (b < 0):
        r -= b
    return r


def product_except_self(nums):
    n = len(nums)
    out = []

    prefix = 1
    for i in range(n):
        out.append(prefix)
        prefix *= nums[i]

    suffix = 1
    for j in range(n - 1, -1, -1):
        out[j] *= suffix
        suffix *= nums[j]

    return out


def lcg_vals(seed, n):
    out = []
    x = seed
    for _ in range(n):
        x = (x * 1103515245 + 12345) % 2147483648
        out.append(1 - 2 * ((x // 65536) % 2))
    return out


def main():
    arrays = [lcg_vals(j + 1, N) for j in range(NP)]

    sink = 0
    for it in range(ITERS):
        idx = (it * 3) % NP
        res = product_except_self(arrays[idx])
        for v, val in enumerate(res):
            sink = trunc_mod(sink + (v + 1) * val, 1000000007)
    print(sink)


if __name__ == "__main__":
    main()

File: leetcode/201-300/238-product-of-array-except-self/bench/product_except_self.rs
// Benchmark harness for LeetCode #238 — Product of Array Except Self.
// Mirrors product_except_self.kara algorithm-for-algorithm.

fn product_except_self(nums: &[i64]) -> Vec<i64> {
    let n = nums.len();
    let mut out: Vec<i64> = Vec::new();

    let mut prefix: i64 = 1;
    for i in 0..n {
        out.push(prefix);
        prefix *= nums[i];
    }

    let mut suffix: i64 = 1;
    let mut j = n as i64 - 1;
    while j >= 0 {
        out[j as usize] *= suffix;
        suffix *= nums[j as usize];
        j -= 1;
    }

    out
}

fn lcg_vals(seed: i64, n: i64) -> Vec<i64> {
    let mut out: Vec<i64> = Vec::new();
    let mut x = seed;
    for _ in 0..n {
        x = (x * 1103515245 + 12345) % 2147483648;
        out.push(1 - 2 * ((x / 65536) % 2));
    }
    out
}

fn main() {
    let np: i64 = 8;
    let n: i64 = 100000;
    let iters: i64 = 400;

    let arrays: Vec<Vec<i64>> = (0..np).map(|j| lcg_vals(j + 1, n)).collect();

    let mut sink: i64 = 0;
    for it in 0..iters {
        let idx = ((it * 3) % np) as usize;
        let res = product_except_self(&arrays[idx]);
        for (v, &val) in res.iter().enumerate() {
            sink = (sink + (v as i64 + 1) * val) % 1000000007;
        }
    }
    println!("{}", sink);
}
